banditpylib/protocols/collaborative_protocol.py

```diff
@@ -0,0 +1,138 @@
+from typing import List, cast, Dict, Tuple
+from copy import deepcopy as dcopy
+
+import numpy as np
+from absl import logging
+
+from banditpylib.bandits import Bandit
+from banditpylib.data_pb2 import Trial, Actions
+from banditpylib.learners import Learner
+from banditpylib.learners.collaborative_learner import CollaborativeBAILearner
+from .utils import Protocol
+
+
+class CollaborativeLearningProtocol(Protocol):
+  """Collaborative learning protocol :cite:'tao2019collaborative'
+
+  This protocol is used to simulate the multi-agent game
+  as discussed in the paper. It runs in rounds. During each round,
+  the protocol runs the following steps in sequence:
+
+  * fetch the state of the environment and ask each learner for actions;
+  * send the actions to the enviroment for execution;
+  * update each learner with the corresponding feedback of the environment;
+  * repeat the above steps until every agent enters the WAIT or STOP state;
+  * if there is at least one agent in WAIT state, then receive information
+    broadcasted from every waiting agent, and use it to decide
+    the parameters of the next round.
+
+  The simulation stopping criteria is:
+
+  * every agent enters STOP state;
+
+  Algorithm is guaranteed to stop before total number of time-steps
+  achieve `horizon`.
+
+  :param Bandit bandit: bandit environment
+  :param List[CollaborativeBAILearner] learners: learners that will be compared
+
+  .. note::
+    During a timestep, a learner may want to perform multiple actions, which is
+    so-called batched learner. In this case, eah action counts as a timestep
+    used.
+  """
+  def __init__(self,
+               bandit: Bandit, learners: List[CollaborativeBAILearner]):
+    super().__init__(bandit=bandit, learners=cast(List[Learner], learners))
+
+  @property
+  def name(self) -> str:
+    return 'collaborative_learning_protocol'
+
+  def _one_trial(self, random_seed: int, debug: bool) -> bytes:
+    if debug:
+      logging.set_verbosity(logging.DEBUG)
+    np.random.seed(random_seed)
+
+    # initialization
+    current_learner = cast(CollaborativeBAILearner, self.current_learner)
+    current_learner.reset()
+    agents = current_learner.get_agents()
+    bandits = []
+    master = current_learner.get_master()
+    for _ in range(len(agents)):
+      bandits.append(dcopy(self.bandit))
+      bandits[-1].reset()
+
+    trial = Trial()
+    trial.bandit = self.bandit.name
+    trial.learner = current_learner.name
+
+    communication_rounds, total_pulls = 0, 0
+    active_agent_ids = list(range(len(agents)))
+    agent_arm_assignment = master.initial_arm_assignment(active_agent_ids)
```

Comment:
initial_arm_assignment accept number of agents.

Reply:
master already has num_agents, so I will pass no arguments